drivers/can/can_native_posix.c

```diff
@@ -0,0 +1,327 @@
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * CANBUS driver for native_posix board. This is meant to test CANBUS
+ * connectivity between host and Zephyr.
+ */
+
+#define DT_DRV_COMPAT zephyr_native_posix_can
+
+#define LOG_MODULE_NAME canbus_posix
+#define LOG_LEVEL CONFIG_CAN_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+
+#include <stdio.h>
+
+#include <kernel.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <stddef.h>
+
+#include <net/net_pkt.h>
+#include <net/net_l2.h>
+#include <net/net_core.h>
+#include <net/net_if.h>
+
+#include <net/can.h>
+#include <net/socket_can.h>
+
+#include "can_native_posix_priv.h"
+#include "socket_can_context.h"
+
+#define NET_BUF_TIMEOUT K_MSEC(100)
+
+static int read_data(struct socket_can_context *ctx, int fd)
+{
+	struct net_pkt *pkt;
+	int count;
+
+	struct can_frame frame;
+
+	count = canbus_np_read_data(fd, (void *)(&frame), sizeof(frame));
+
+	if (count <= 0) {
+		return 0;
+	}
+
+	struct zcan_frame zframe;
+
+	can_copy_frame_to_zframe(&frame, &zframe);
+
+	pkt = net_pkt_rx_alloc_with_buffer(ctx->iface, sizeof(zframe), AF_CAN,
+					   0, NET_BUF_TIMEOUT);
+	if (!pkt) {
+		return -ENOMEM;
+	}
+
+	if (net_pkt_write(pkt, (void *)(&zframe), sizeof(zframe))) {
+		net_pkt_unref(pkt);
+		return -ENOBUFS;
+	}
+
+	if (net_recv_data(ctx->iface, pkt) < 0) {
+		net_pkt_unref(pkt);
+	}
+
+	return 0;
+}
+
+static void canbus_np_rx(struct socket_can_context *ctx)
+{
+	LOG_DBG("Starting ZCAN RX thread");
+
+	while (1) {
+		if (ctx->iface && net_if_is_up(ctx->iface)) {
+			while (!canbus_np_wait_data(ctx->dev_fd)) {
+				read_data(ctx, ctx->dev_fd);
+			}
+		}
+
+		k_sleep(K_MSEC(10));
+	}
+}
+
+static int canbus_np_send(const struct device *dev,
+			  const struct zcan_frame *msg, k_timeout_t timeout,
+			  can_tx_callback_t callback_isr, void *callback_arg)
+{
+	struct socket_can_context *ctx = dev->data;
+	int ret = -ENODEV;
+
+	ARG_UNUSED(timeout);
+	ARG_UNUSED(callback_isr);
```

Comment:
I would rather return an error in case someone pass a callback.
Since callbacks are not supported, this driver may not work with other CAN subsys like ISO-TP or canopen.